src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,4058 @@
+        use_text_encoder: Optional[bool] = None,
+        use_speech_encoder: Optional[bool] = None,
+    ):
+        super().__init__(config)
+
+        use_text_encoder = use_text_encoder if use_text_encoder is not None else config.use_text_encoder
+        use_speech_encoder = use_speech_encoder if use_speech_encoder is not None else config.use_speech_encoder
+
+        if not use_text_encoder and not use_speech_encoder:
+            raise ValueError(
+                "`SeamlessM4TMultiModalToTextModel` can't be used without a speech encoder or a text encoder. You should have either `use_text_encoder=True` or `use_speech_encoder=True`."
+            )
+
+        if use_text_encoder:
+            self.text_encoder = SeamlessM4TEncoder(config)
+
+        if use_speech_encoder:
+            self.speech_encoder = SeamlessM4TSpeechEncoder(config)
+
+        self.text_decoder = SeamlessM4TDecoder(config)
+
+        self.post_init()
+
+    def get_decoder(self):
+        return self.text_decoder
+
+    def get_encoder(self):
+        return self.text_encoder
+
+    # priority is given to None input values
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        input_values: Optional[torch.FloatTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        decoder_input_ids: Optional[torch.LongTensor] = None,
+        decoder_attention_mask: Optional[torch.LongTensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        decoder_head_mask: Optional[torch.Tensor] = None,
+        cross_attn_head_mask: Optional[torch.Tensor] = None,
+        encoder_outputs: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        decoder_inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        **kwargs,
+    ) -> Union[Seq2SeqModelOutput, Tuple[torch.FloatTensor]]:
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if encoder_outputs is None and input_values is not None:
+            if input_ids is not None:
+                logger.warning(
+                    "`input_ids` is not `None` but `input_values` has been given. `input_values` will be used in priority through the `speech_encoder`. Make sure that `input_values` and `input_ids` are mutually exclusive."
```

Comment:
Not a huge fan of this priority method - think we can explicitly define the inputs for each of the task-specific classes, and then handle this for the TTS model where we might have both

(I would probably error out here rather than throw a warning as well)